### PR TITLE
108 test deform

### LINKED ISSUE #108: Test + docs: label/deform.py
# Issue description

- Deform.py: 1 function, local displacement energy that could be refactored into a class or separated into individual unit functions to facilitate testing. No docstrings, requires some documentation updates but code semi understandable.

## Type of Issue

- [x] Documentation update
- [x] Test request
- [x] Linting request

#### What's the expected result?

Unit tests for deform.py. Considering refactoring the function into submodules to facilitate testing.

### PR DESCRIPTION
**Pull request by QianyeYang**
# Description

Please include a summary of the change and which issue is fixed. Please also include
relevant motivation and context. List any dependencies that are required for this
change.

Fixes #108 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not
      work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so
we can reproduce. Please also list any relevant details for your test configuration If
there are no tests written, flag with label so we can help generate some tests before
merge.

- [ ] Test A
- [ ] Test B

**Test Configuration**:

- Firmware version:
- Hardware:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: deepreg/model/loss/deform.py [hunk 1]
```diff
@@ -1,49 +1,55 @@
 import tensorflow as tf
```
Comment (NMontanaBrown) [inline deepreg/model/loss/deform.py:1]: Module docstring missing: can you add this to clarify what the module does?

File: deepreg/model/loss/deform.py [hunk 2]
```diff
@@ -1,49 +1,55 @@
...
-        return tf.reduce_mean(
-            dTdxx ** 2
-            + dTdyy ** 2
-            + dTdzz ** 2
-            + 2 * dTdxy ** 2
-            + 2 * dTdxz ** 2
-            + 2 * dTdyz ** 2,
-            [1, 2, 3, 4],
-        )
+def gradient_dx(fv):
```
Comment (NMontanaBrown) [inline deepreg/model/loss/deform.py:4]: Function docstring missing

File: deepreg/model/loss/deform.py [hunk 3]
```diff
@@ -1,49 +1,55 @@
...
+    return (fv[:, 2:, 1:-1, 1:-1] - fv[:, :-2, 1:-1, 1:-1]) / 2
+
+
+def gradient_dy(fv):
```
Comment (NMontanaBrown) [inline deepreg/model/loss/deform.py:8]: Function docstring missing

File: deepreg/model/loss/deform.py [hunk 4]
```diff
@@ -1,49 +1,55 @@
...
+    return (fv[:, 1:-1, 2:, 1:-1] - fv[:, 1:-1, :-2, 1:-1]) / 2
+
+
+def gradient_dz(fv):
```
Comment (NMontanaBrown) [inline deepreg/model/loss/deform.py:12]: function docstring missing

File: deepreg/model/loss/deform.py [hunk 5]
```diff
@@ -1,49 +1,55 @@
...
+    return (fv[:, 1:-1, 1:-1, 2:] - fv[:, 1:-1, 1:-1, :-2]) / 2
+
 
+def gradient_txyz(Txyz, fn):
```
Comment (NMontanaBrown) [inline deepreg/model/loss/deform.py:16]: Function docstring missing

File: deepreg/model/loss/deform.py [hunk 6]
```diff
@@ -1,49 +1,55 @@
...
+    return tf.stack([fn(Txyz[..., i]) for i in [0, 1, 2]], axis=4)
+
+
+def compute_gradient_norm(displacement, l1=False):
```
Comment (NMontanaBrown) [inline deepreg/model/loss/deform.py:20]: Function docstring missing

File: deepreg/model/loss/deform.py [hunk 7]
```diff
@@ -1,49 +1,55 @@
...
+    dTdy = gradient_txyz(displacement, gradient_dy)
+    dTdz = gradient_txyz(displacement, gradient_dz)
+    if l1:
+        norms = tf.abs(dTdx) + tf.abs(dTdy) + tf.abs(dTdz)
+    else:
+        norms = dTdx ** 2 + dTdy ** 2 + dTdz ** 2
+    return tf.reduce_mean(norms, [1, 2, 3, 4])
+
+
+def compute_bending_energy(displacement):
```
Comment (NMontanaBrown) [inline deepreg/model/loss/deform.py:31]: Function docstring missing

File: deepreg/model/loss/deform.py [hunk 8]
```diff
@@ -1,49 +1,55 @@
...
+        + dTdyy ** 2
+        + dTdzz ** 2
+        + 2 * dTdxy ** 2
+        + 2 * dTdxz ** 2
+        + 2 * dTdyz ** 2,
+        [1, 2, 3, 4],
+    )
+
+
+def local_displacement_energy(ddf, energy_type, **kwargs):
```
Comment (NMontanaBrown) [inline deepreg/model/loss/deform.py:52]: Line 60 not tested: (raise Value Error). Can you add a pytest assertion test here for completeness?

Comment (QianyeYang) [reply deepreg/model/loss/deform.py:99]: Sorry for the delay, I've added the doc strings and assertion test.
